File: frontend/src/services/RevenueCatService.ts
// Note: Platform import removed as this is a web-only implementation

export interface RevenueCatConfig {
  apiKey: string;
  baseUrl: string;
}

export interface PurchaseRequest {
  productId: string;
  userId: string;
  userEmail: string;
  planId: string;
  planName: string;
  price: number;
  currency: string;
}

export interface PurchaseResponse {
  success: boolean;
  transactionId?: string;
  subscriptionId?: string;
  error?: string;
  redirectUrl?: string;
}

export interface SubscriptionStatus {
  isActive: boolean;
  planId?: string;
  planName?: string;
  expiresAt?: string;
  isTrial?: boolean;
  trialEndsAt?: string;
}

class RevenueCatService {
  private config: RevenueCatConfig;
  private isInitialized = false;

  constructor() {
    this.config = {
      apiKey: import.meta.env.VITE_REVENUECAT_API_KEY || '',
      baseUrl: import.meta.env.VITE_REVENUECAT_BASE_URL || 'https://api.revenuecat.com/v1'
    };
  }

  async initialize(): Promise<boolean> {
    if (!this.config.apiKey) {
      console.error('RevenueCat API key not configured');
      return false;
    }

    this.isInitialized = true;
    return true;
  }

  async createCustomer(userId: string, userEmail: string): Promise<boolean> {
    if (!this.isInitialized) {
      await this.initialize();
    }

    try {
      const response = await fetch(`${this.config.baseUrl}/subscribers/${userId}`, {
        method: 'POST',
        headers: {
          'Authorization': `Bearer ${this.config.apiKey}`,
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          app_user_id: userId,
          email: userEmail
        })
      });

      return response.ok;
    } catch (error) {
      console.error('Error creating RevenueCat customer:', error);
      return false;
    }
  }

  async initiatePurchase(purchaseRequest: PurchaseRequest): Promise<PurchaseResponse> {
    if (!this.isInitialized) {
      await this.initialize();
    }

    try {
      // Create customer first
      const customerCreated = await this.createCustomer(
        purchaseRequest.userId,
        purchaseRequest.userEmail
      );

      if (!customerCreated) {
        return {
          success: false,
          error: 'Failed to create customer account'
        };
      }

      // For web integration, we'll redirect to RevenueCat's hosted checkout
      const checkoutUrl = this.generateCheckoutUrl(purchaseRequest);
      
      return {
        success: true,
        redirectUrl: checkoutUrl
      };
    } catch (error) {
      console.error('Error initiating purchase:', error);
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error occurred'
      };
    }
  }

  private generateCheckoutUrl(purchaseRequest: PurchaseRequest): string {
    const params = new URLSearchParams({
      app_user_id: purchaseRequest.userId,
      email: purchaseRequest.userEmail,
      product_id: purchaseRequest.productId,
      plan_name: purchaseRequest.planName,
      price: purchaseRequest.price.toString(),
      currency: purchaseRequest.currency,
      return_url: `${window.location.origin}/payment/success`,
      cancel_url: `${window.location.origin}/join?step=pricing`
    });

    return `${this.config.baseUrl}/checkout?${params.toString()}`;
  }

  async getSubscriptionStatus(userId: string): Promise<SubscriptionStatus> {
    if (!this.isInitialized) {
      await this.initialize();
    }

    try {
      const response = await fetch(`${this.config.baseUrl}/subscribers/${userId}`, {
        headers: {
          'Authorization': `Bearer ${this.config.apiKey}`,
          'Content-Type': 'application/json'
        }
      });

      if (!response.ok) {
        return { isActive: false };
      }

      const data = await response.json();
      const subscriptions = data.subscriber?.subscriptions || {};
      
      // Find the first active subscription
      const activeSubscription = Object.values(subscriptions).find((sub: any) => sub.is_active);
      const subscription = activeSubscription || null;

      return {
        isActive: subscription?.is_active || false,
        planId: subscription?.product_id,
        planName: subscription?.product_name,
        expiresAt: subscription?.expires_date,
        isTrial: subscription?.is_trial_period || false,
        trialEndsAt: subscription?.trial_end_date
      };
    } catch (error) {
      console.error('Error getting subscription status:', error);
      return { isActive: false };
    }
  }

  async handleWebhook(webhookData: any): Promise<boolean> {
    // Handle RevenueCat webhooks for subscription updates
    try {
      // Verify webhook signature if needed
      // Process subscription events (created, renewed, cancelled, etc.)
      
      console.log('Processing RevenueCat webhook:', webhookData);
      return true;
    } catch (error) {
      console.error('Error processing webhook:', error);
      return false;
    }
  }
}

export const revenueCatService = new RevenueCatService();
export default revenueCatService;

File: backend/services/revenuecat_service.py
import httpx
import json
from typing import Optional, Dict, Any
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, update
from models.user import User
from models.business import Business
from core.config import settings
import logging

logger = logging.getLogger(__name__)

class RevenueCatService:
    """Service for handling RevenueCat integration"""
    
    def __init__(self):
        self.api_key = settings.REVENUECAT_API_KEY
        self.base_url = settings.REVENUECAT_BASE_URL
        self.headers = {
            'Authorization': f'Bearer {self.api_key}',
            'Content-Type': 'application/json'
        }
    
    async def create_customer(self, user_id: str, email: str) -> bool:
        """Create a customer in RevenueCat"""
        try:
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    f"{self.base_url}/subscribers/{user_id}",
                    headers=self.headers,
                    json={
                        "app_user_id": user_id,
                        "email": email
                    }
                )
                return response.status_code in [200, 201]
        except Exception as e:
            logger.error(f"Error creating RevenueCat customer: {e}")
            return False
    
    async def get_subscription_status(self, user_id: str) -> Optional[Dict[str, Any]]:
        """Get subscription status from RevenueCat"""
        try:
            async with httpx.AsyncClient() as client:
                response = await client.get(
                    f"{self.base_url}/subscribers/{user_id}",
                    headers=self.headers
                )
                
                if response.status_code == 200:
                    data = response.json()
                    subscriber = data.get('subscriber', {})
                    subscriptions = subscriber.get('subscriptions', {})
                    
                    # Find active subscription
                    for product_id, subscription in subscriptions.items():
                        if subscription.get('is_active', False):
                            return {
                                'is_active': True,
                                'product_id': product_id,
                                'expires_date': subscription.get('expires_date'),
                                'is_trial_period': subscription.get('is_trial_period', False),
                                'trial_end_date': subscription.get('trial_end_date'),
                                'period_type': subscription.get('period_type'),
                                'store': subscription.get('store')
                            }
                    
                    return {'is_active': False}
                else:
                    logger.error(f"Error getting subscription status: {response.status_code}")
                    return None
        except Exception as e:
            logger.error(f"Error getting subscription status: {e}")
            return None
    
    async def handle_webhook(self, webhook_data: Dict[str, Any], db: AsyncSession) -> bool:
        """Handle RevenueCat webhook events"""
        try:
            event_type = webhook_data.get('type')
            subscriber = webhook_data.get('subscriber', {})
            app_user_id = subscriber.get('app_user_id')
            
            if not app_user_id:
                logger.error("No app_user_id in webhook data")
                return False
            
            # Get user from database
            result = await db.execute(select(User).where(User.id == int(app_user_id)))
            user = result.scalar_one_or_none()
            
            if not user:
                logger.error(f"User not found for app_user_id: {app_user_id}")
                return False
            
            if event_type == 'INITIAL_PURCHASE':
                await self._handle_initial_purchase(webhook_data, user, db)
            elif event_type == 'RENEWAL':
                await self._handle_renewal(webhook_data, user, db)
            elif event_type == 'CANCELLATION':
                await self._handle_cancellation(webhook_data, user, db)
            elif event_type == 'NON_RENEWING_PURCHASE':
                await self._handle_non_renewing_purchase(webhook_data, user, db)
            elif event_type == 'PRODUCT_CHANGE':
                await self._handle_product_change(webhook_data, user, db)
            
            await db.commit()
            return True
            
        except Exception as e:
            logger.error(f"Error handling webhook: {e}")
            await db.rollback()
            return False
    
    async def _handle_initial_purchase(self, webhook_data: Dict[str, Any], user: User, db: AsyncSession):
        """Handle initial purchase event"""
        logger.info(f"Handling initial purchase for user {user.id}")
        # Update user subscription status
        # You can add subscription fields to User model if needed
        pass
    
    async def _handle_renewal(self, webhook_data: Dict[str, Any], user: User, db: AsyncSession):
        """Handle subscription renewal event"""
        logger.info(f"Handling renewal for user {user.id}")
        # Update subscription status
        pass
    
    async def _handle_cancellation(self, webhook_data: Dict[str, Any], user: User, db: AsyncSession):
        """Handle subscription cancellation event"""
        logger.info(f"Handling cancellation for user {user.id}")
        # Update subscription status
        pass
    
    async def _handle_non_renewing_purchase(self, webhook_data: Dict[str, Any], user: User, db: AsyncSession):
        """Handle non-renewing purchase event"""
        logger.info(f"Handling non-renewing purchase for user {user.id}")
        # Update subscription status
        pass
    
    async def _handle_product_change(self, webhook_data: Dict[str, Any], user: User, db: AsyncSession):
        """Handle product change event"""
        logger.info(f"Handling product change for user {user.id}")
        # Update subscription plan
        pass
    
    def generate_checkout_url(self, user_id: str, email: str, product_id: str, plan_name: str, price: float, currency: str = 'EUR') -> str:
        """Generate RevenueCat checkout URL for web integration"""
        params = {
            'app_user_id': user_id,
            'email': email,
            'product_id': product_id,
            'plan_name': plan_name,
            'price': str(price),
            'currency': currency,
            'return_url': f"{settings.BASE_URL}/payment/success",
            'cancel_url': f"{settings.BASE_URL}/join?step=pricing"
        }
        
        query_string = '&'.join([f"{k}={v}" for k, v in params.items()])
        return f"{self.base_url}/checkout?{query_string}"

# Global instance
revenuecat_service = RevenueCatService()
